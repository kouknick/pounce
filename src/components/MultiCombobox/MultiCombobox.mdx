A thing to remember is that all the comboboxes have a controlled state when
it comes to the selected items. Thus, they do not maintain their own internal state
but expect you to pass them explicitly the selected items.

You get notified of changes to the selections through the `onChange` prop.

---

A MultiCombobox can be basic:

```typescript jsx


const Example3 = () => {
  const [selectedItems, updateSelectedItems] = React.useState([]);
  return (
    <MultiCombobox
      label="Choose a car manufacturer"
      items={['Toyota', 'Ford', 'Chevrolet', 'BMW', 'Mercedes', 'Hammer', 'Dodge', 'Audi']}
      onChange={updateSelectedItems}
      value={selectedItems}
      inputProps={{ placeholder: 'Search for a manufacturer' }}
      rootProps={{ width: 400 }}
    />
  );
};

<Example3 />;
```

A MultiCombobox have search functionality:

```typescript jsx


const Example4 = () => {
  const [selectedItems, updateSelectedItems] = React.useState([]);
  return (
    <MultiCombobox
      searchable
      label="Choose a car manufacturer"
      items={['Toyota', 'Ford', 'Chevrolet', 'BMW', 'Mercedes', 'Hammer', 'Dodge', 'Audi']}
      onChange={updateSelectedItems}
      value={selectedItems}
      inputProps={{ placeholder: 'Search for a manufacturer' }}
      rootProps={{ width: 400 }}
    />
  );
};

<Example4 />;
```

A MultiCombobox can allow user to add values of their own:

```typescript jsx


const Example4 = () => {
  const [selectedItems, updateSelectedItems] = React.useState([]);
  return (
    <MultiCombobox
      allowAdditions
      searchable
      label="Choose a car manufacturer"
      items={[]}
      onChange={updateSelectedItems}
      value={selectedItems}
      inputProps={{ placeholder: 'Search for a manufacturer' }}
      rootProps={{ width: 400 }}
    />
  );
};

<Example4 />;
```

A MultiCombobox can allow user to only add values that satisfy a certain condition:

```typescript jsx


const Example4 = () => {
  const [selectedItems, updateSelectedItems] = React.useState([]);
  return (
    <MultiCombobox
      allowAdditions
      searchable
      label="Only words that contain the word Panther"
      items={[]}
      onChange={updateSelectedItems}
      validateAddition={text => text.includes('Panther')}
      value={selectedItems}
      inputProps={{ placeholder: 'Must contain the word Panther' }}
      rootProps={{ width: 400 }}
    />
  );
};

<Example4 />;
```
