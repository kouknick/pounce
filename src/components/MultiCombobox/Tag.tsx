import Icon from '../Icon';
import Flex, { FlexProps } from '../Flex';
import React from 'react';
import AbstractButton from '../AbstractButton';

export interface TagProps extends FlexProps {
  /**
   * What happens when the "X" button is pressed. No "X" button is rendered, if this prop is not
   * specified
   * */
  onRemove?: () => void;
}

/** A chip is an entry in a combobox, but can be used anywhere */
const Tag: React.FC<TagProps> = ({ children, onRemove, ...rest }) => (
  <Flex
    bg="blue-600"
    borderRadius="pill"
    cursor="default"
    align="center"
    px={2}
    fontSize="small"
    fontWeight="medium"
    color="gray-50"
    zIndex={1}
    {...rest}
  >
    {children}
    <AbstractButton onClick={onRemove} ml={2} aria-label="Remove">
      <Icon size="x-small" type="close" />
    </AbstractButton>
  </Flex>
);

export default Tag;
