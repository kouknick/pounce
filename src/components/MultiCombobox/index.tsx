export { default } from './MultiCombobox';
export * from './MultiCombobox';
