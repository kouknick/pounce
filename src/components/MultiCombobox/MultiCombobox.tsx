/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import Downshift, { DownshiftState, StateChangeOptions } from 'downshift';
import { filter as fuzzySearch } from 'fuzzaldrin';
import Box from '../Box';
import MenuItem from '../utils/MenuItem';
import Icon from '../Icon';
import Flex from '../Flex';
import { InputControl, InputLabel, InputElement, InputElementProps } from '../utils/Input';
import Tag from './Tag';
import { typedMemo } from '../../utils/helpers';
import Menu from '../utils/Menu';

export type MultiComboboxProps<T> = {
  /** Callback when the selection changes */
  onChange: (value: T[]) => void;

  /** The label associated with this dropdown form element */
  label: string;

  /** The variant of the component that decides the colors */
  variant?: 'solid' | 'outline';

  /** Whether the label should get visually hidden */
  hideLabel?: boolean;

  /** A list of entries that the dropdown will have as options */
  items: T[];

  /**
   * A function that converts the an item to a string. This is the value that the dropdown will
   * expose to the user. This is also the value that the dropdown will use internally for comparisons
   * so it should be unique (which should be by default since you wouldn't want to expose duplicate
   * values to the user) */
  itemToString?: (item: T) => string;

  /**
   * A function that accepts an item as a parameteer and returns `true` if the item should be
   * disabled or `false` otherwise. Defaults to `() => false`.
   */
  disableItem?: (item: T) => boolean;

  /**
   * The value of the item that is currently selected. The component is a controlled one,
   * so the the selected value should be provided explicitly to the dropdown
   * */
  value: T[];

  /** Whether the multi-combobox has an invalid value */
  invalid?: boolean;

  /** Whether the multi-combobox is required or not */
  required?: boolean;

  /** Whether the multi-combobox is disabled or not */
  disabled?: boolean;

  /** A placeholder that's visible when the user focuses on the Combobox */
  placeholder?: string;

  /** Whether the MultiCombobox should have an input to search results */
  searchable?: boolean;

  /** @ignore */
  hidden?: boolean;

  /**
   * Allow the user to add custom entries to the dropdown instead of limiting selections to the
   * predefined set of options. The `searchable` prop should be true in order for
   * this functionality to work.
   * */
  allowAdditions?: boolean;

  /**
   * The maximum number of results that the MultiCombobox should show. Default value is
   * `undefined` to display which displays all of them.
   * */
  maxResults?: number;

  /** The maximum height (in pixels) of the MultiCombobox dropdown. Defaults to 300. */
  maxHeight?: number;

  /**
   * A function that runs before a custom item is added by the user. If it returns `true`, then this
   * item will be added to the selection. If not, then this item won't be added
   * */
  validateAddition?: (userEnteredInput: string) => boolean;
};

const stateReducer = (state: DownshiftState<any>, changes: StateChangeOptions<any>) => {
  switch (changes.type) {
    case Downshift.stateChangeTypes.keyDownEnter:
    case Downshift.stateChangeTypes.clickItem:
      // Whenever the user makes a selection, make sure to reset the input so that he can keep
      // searching for more. We also make sure to always highlight the first item
      return {
        ...changes,
        inputValue: '',
        highlightedIndex: state.highlightedIndex,
        isOpen: true,
      };
    case Downshift.stateChangeTypes.blurInput:
      return {
        ...changes,
        inputValue: '',
      };

    default:
      return changes;
  }
};

/**
 * A simple MultiCombobox can be thought of as a typical `<select>` component. Whenerever you would
 * use a normal select, you should now pass the `<MultiCombobox>` component.
 */
function MultiCombobox<Item>({
  onChange,
  value,
  variant = 'outline',
  items,
  disableItem = () => false,
  searchable = false,
  label = '',
  hideLabel,
  disabled = false,
  placeholder = '',
  itemToString = item => String(item),
  allowAdditions = false,
  validateAddition = () => true,
  maxHeight = 300,
  maxResults,
  invalid,
  hidden,
  ...rest
}: MultiComboboxProps<Item>): React.ReactElement<MultiComboboxProps<Item>> {
  const getVariant = React.useCallback(
    isOpen => {
      if (variant === 'solid') {
        return 'solid';
      }
      return isOpen ? 'solid' : 'outline';
    },
    [variant]
  );
  const removeItem = (item: any) => {
    onChange(value.filter(i => i !== item));
  };

  const addSelectedItem = (item: any) => {
    if (item !== null) {
      onChange([...value, item]);
    }
  };

  const itemsPt = hideLabel ? 3 : '19px';

  return (
    <Downshift<Item>
      stateReducer={stateReducer}
      onChange={addSelectedItem}
      selectedItem={null}
      itemToString={item => (item ? itemToString(item) : '')}
      initialInputValue=""
    >
      {({
        getRootProps,
        getInputProps,
        getItemProps,
        getMenuProps,
        getLabelProps,
        inputValue,
        isOpen,
        toggleMenu,
        openMenu,
        selectedItem,
        selectItem,
      }) => {
        const multiComboboxVariant = getVariant(isOpen);
        // If it's a multicombobox we DON'T WANT t oinclude the results already selected and also
        // we want to make sure that the results get filtered by the search term of the user
        const nonSelectedItems = items.filter(
          item => !value.map(itemToString).includes(itemToString(item))
        );

        // From the non-selected items, make sure to filter the ones that match the user's
        // search term. To do that we convert our items to their string representations
        const strResults = fuzzySearch(nonSelectedItems.map(itemToString), inputValue || '');

        // and then convert those strings back to the original shape of the items, while making
        // sure to only display a (potentially) limited number of them
        const results = items
          .filter(item => strResults.includes(itemToString(item)))
          .slice(0, maxResults);

        // Only show the items that have not been selected

        // We add 2 types of additional data to the input that is going to be renders:
        // 1. A handler for the `Delete` button, so that you can delete tokens with a single key
        // 3. When the combobox is not searchable, we make the input "behave" like a div. We
        // still want an input though for placeholder, spacings, etc.
        const additionalInputProps = {
          ...rest,
          ...(!searchable && {
            cursor: 'pointer',
            onMouseDown: toggleMenu,
            readOnly: true,
            placeholder: !value.length ? placeholder : '',
            position: !value.length ? 'static' : 'absolute',
          }),
          ...(searchable && {
            placeholder,
            mt: (isOpen && value.length) || (isOpen && value.length && allowAdditions) ? -4 : 0,
            position: isOpen || (isOpen && value.length && allowAdditions) ? 'static' : 'absolute',
          }),
          width: '100%',
          height: '100%',
          top: 0,
          left: 0,
          onFocus: openMenu,
          onKeyDown: (event: React.KeyboardEvent) => {
            // Allow deletions of selections by pressing backspace
            if (event.key === 'Backspace' && !inputValue) {
              removeItem(value[value.length - 1]);
            }

            // Allow the user to add custom selections if both `searchable` and `allowAdditions`
            // have a truthy value
            const trimmedInputValue = (inputValue || '').trim();
            if (
              (event.key === 'Enter' || event.key === ',') &&
              allowAdditions &&
              trimmedInputValue
            ) {
              event.preventDefault();

              // By default validateAddition always returns true. Can be overriden by the user
              // for fine-grained addition
              if (validateAddition && validateAddition(trimmedInputValue)) {
                selectItem((inputValue as unknown) as Item, { inputValue: '', isOpen: true });
              }
            }
          },
          onBlur: () => {
            const trimmedInputValue = (inputValue || '').trim();
            if (allowAdditions && trimmedInputValue) {
              selectItem((trimmedInputValue as unknown) as Item, { inputValue: '' });
            }
          },
          onPaste: (e: React.ClipboardEvent<HTMLInputElement>) => {
            // prevent it when we can only select values from the dropdown
            if (!allowAdditions) {
              return;
            }

            // Get clipboard data and split them based on newline and/or commas
            const clipboardData = e.clipboardData.getData('Text');
            const items = (clipboardData
              .replace(/\r?\n/g, ',')
              .split(',')
              .map(str => str.trim()) as unknown) as Item[];

            if (items.length > 1) {
              // Prevent the text from actually being pasted to the underlying input
              e.preventDefault();
              e.stopPropagation();

              // extend existing values with new ones
              onChange([...value, ...items]);
            }
          },
        };

        return (
          <Box position="relative" {...getRootProps()}>
            <Box position="relative">
              <InputControl
                invalid={invalid}
                disabled={disabled}
                variant={multiComboboxVariant}
                hidden={hidden}
              >
                {value.length > 0 && (
                  <Flex as="ul" wrap="wrap" pl={3} pr={10} pt={itemsPt} pb="2px">
                    {value.map(selectedItem => (
                      <Tag
                        as="li"
                        key={itemToString(selectedItem)}
                        m={1}
                        onRemove={() => removeItem(selectedItem)}
                      >
                        {itemToString(selectedItem)}
                      </Tag>
                    ))}
                  </Flex>
                )}
                <InputElement
                  as="input"
                  type="text"
                  standalone={hideLabel}
                  {...(getInputProps(additionalInputProps) as Omit<InputElementProps, 'ref'>)}
                />
                <InputLabel visuallyHidden={hideLabel} raised={!!value.length} {...getLabelProps()}>
                  {label}
                </InputLabel>
              </InputControl>

              {items.length > 0 && (
                <Icon
                  type={isOpen ? 'caret-up' : 'caret-down'}
                  position="absolute"
                  pointerEvents="none"
                  my="auto"
                  top={0}
                  bottom={0}
                  right={4}
                />
              )}
            </Box>
            <Menu
              as="ul"
              maxHeight={maxHeight}
              isOpen={isOpen && results.length > 0}
              {...getMenuProps()}
            >
              {results.map(item => (
                <MenuItem
                  {...getItemProps({ item, disabled: disableItem(item) })}
                  as="li"
                  listStyle="none"
                  key={itemToString(item)}
                  selected={item === selectedItem}
                >
                  {itemToString(item)}
                </MenuItem>
              ))}
            </Menu>
          </Box>
        );
      }}
    </Downshift>
  );
}

export default typedMemo(MultiCombobox);
