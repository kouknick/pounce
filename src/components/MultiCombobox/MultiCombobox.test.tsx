import { renderWithTheme, fireClickAndMouseEvents, fireEvent } from 'test-utils';
import React from 'react';
import MultiCombobox, { MultiComboboxProps } from './index';

const items = ['Toyota', 'Ford', 'Chevrolet', 'BMW', 'Mercedes', 'Hammer', 'Dodge', 'Audi'];

const ControlledMultiCombobox = <T,>(props: Partial<MultiComboboxProps<T>>) => {
  const [selectedManufacturer, updateSelectedManufacturer] = React.useState<T[]>([]);

  return (
    <MultiCombobox<T>
      label="Choose a car manufacturer"
      onChange={updateSelectedManufacturer}
      value={selectedManufacturer}
      placeholder="Select manufacturers"
      items={props.items}
      {...props}
    />
  );
};

describe('MultiCombobox', () => {
  it('renders', () => {
    const { container } = renderWithTheme(<ControlledMultiCombobox items={items} />);
    expect(container).toMatchSnapshot();
  });

  it('renders with variant="solid"', () => {
    const { container } = renderWithTheme(
      <ControlledMultiCombobox items={items} variant="solid" />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders with clear all option', async () => {
    const { container, getByText, getByPlaceholderText } = renderWithTheme(
      <ControlledMultiCombobox items={items} canClearAllAfter={2} />
    );
    fireClickAndMouseEvents(getByPlaceholderText('Select manufacturers'));
    fireClickAndMouseEvents(await getByText('Toyota'));
    expect(container).toMatchSnapshot();
    fireClickAndMouseEvents(await getByText('Ford'));
    expect(await getByText('Clear all')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('works without being searchable', async () => {
    const { getByText, getByPlaceholderText, getAllByRole } = renderWithTheme(
      <ControlledMultiCombobox items={items} />
    );

    fireClickAndMouseEvents(getByPlaceholderText('Select manufacturers'));
    fireClickAndMouseEvents(await getByText('Toyota'));
    fireClickAndMouseEvents(await getByText('Ford'));

    expect(getAllByRole('tag')).toHaveLength(2);
  });

  it('works while searchable', async () => {
    const { getByText, getByPlaceholderText, getAllByRole } = renderWithTheme(
      <ControlledMultiCombobox items={items} searchable />
    );

    const input = getByPlaceholderText('Select manufacturers');
    fireClickAndMouseEvents(getByPlaceholderText('Select manufacturers'));

    fireEvent.change(input, { target: { value: 'Toyo' } });
    expect(getAllByRole('option')).toHaveLength(1);
    fireClickAndMouseEvents(await getByText('Toyota'));

    fireEvent.change(input, { target: { value: 'Fo' } });
    expect(getAllByRole('option')).toHaveLength(1);
    fireClickAndMouseEvents(await getByText('Ford'));

    expect(getAllByRole('tag')).toHaveLength(2);
  });

  it('allows users to clear all items when selected items are 4 or more', async () => {
    const {
      getByText,
      getByPlaceholderText,
      getAllByRole,
      queryByText,
      queryAllByRole,
    } = renderWithTheme(<ControlledMultiCombobox items={items} canClearAllAfter={4} />);

    fireClickAndMouseEvents(getByPlaceholderText('Select manufacturers'));
    fireClickAndMouseEvents(await getByText('Toyota'));
    fireClickAndMouseEvents(await getByText('Ford'));
    fireClickAndMouseEvents(await getByText('Mercedes'));
    expect(await queryByText('Clear all')).not.toBeInTheDocument();
    fireClickAndMouseEvents(await getByText('Chevrolet'));
    expect(await queryByText('Clear all')).toBeInTheDocument();
    fireClickAndMouseEvents(await getByText('Dodge'));
    expect(getAllByRole('tag')).toHaveLength(5);
    fireClickAndMouseEvents(await getByText('Clear all'));
    expect(queryAllByRole('tag')).toHaveLength(0);
  });

  it("it doesn't allow custom additions by default", async () => {
    const { getByPlaceholderText, queryAllByRole } = renderWithTheme(
      <ControlledMultiCombobox items={items} searchable />
    );

    const input = getByPlaceholderText('Select manufacturers');
    fireClickAndMouseEvents(getByPlaceholderText('Select manufacturers'));

    fireEvent.change(input, { target: { value: 'Random Value' } });
    fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' });
    expect(queryAllByRole('tag')).toHaveLength(0);
  });

  it('works while allowing custom values', async () => {
    const { getByPlaceholderText, getAllByRole } = renderWithTheme(
      <ControlledMultiCombobox items={items} searchable allowAdditions />
    );

    const input = getByPlaceholderText('Select manufacturers');
    fireClickAndMouseEvents(getByPlaceholderText('Select manufacturers'));

    fireEvent.change(input, { target: { value: 'Random Value 1' } });
    fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' });

    fireEvent.change(input, { target: { value: 'Random Value 2' } });
    fireEvent.keyDown(input, { key: ',', code: 'Comma' });
    expect(getAllByRole('tag')).toHaveLength(2);
  });

  it('correctly filters custom values when specified', async () => {
    const { getByPlaceholderText, queryAllByRole } = renderWithTheme(
      <ControlledMultiCombobox
        items={items}
        searchable
        allowAdditions
        validateAddition={text => text.includes('pounce')}
      />
    );

    const input = getByPlaceholderText('Select manufacturers');
    fireClickAndMouseEvents(getByPlaceholderText('Select manufacturers'));

    fireEvent.change(input, { target: { value: 'Random Value' } });
    fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' });
    expect(queryAllByRole('tag')).toHaveLength(0);

    fireEvent.change(input, { target: { value: 'pounce' } });
    fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' });
    expect(queryAllByRole('tag')).toHaveLength(1);
  });

  it("doesn't tokenize value on blur when custom values are disallowed", async () => {
    const { getByPlaceholderText, queryAllByRole } = renderWithTheme(
      <ControlledMultiCombobox items={items} searchable />
    );

    const input = getByPlaceholderText('Select manufacturers') as HTMLInputElement;
    fireClickAndMouseEvents(getByPlaceholderText('Select manufacturers'));

    fireEvent.change(input, { target: { value: 'Toyo' } });
    fireEvent.blur(input);

    expect(queryAllByRole('tag')).toHaveLength(0);
  });

  it('tokenizes value on blur when custom values are allowed', async () => {
    const { getByPlaceholderText, queryAllByRole } = renderWithTheme(
      <ControlledMultiCombobox items={items} searchable allowAdditions />
    );

    const input = getByPlaceholderText('Select manufacturers');
    fireClickAndMouseEvents(getByPlaceholderText('Select manufacturers'));

    fireEvent.change(input, { target: { value: 'Random Value' } });
    fireEvent.blur(input);
    expect(queryAllByRole('tag')).toHaveLength(1);
  });

  it('does not auto-tokenize a single pasted value', async () => {
    const { getByPlaceholderText, queryAllByRole } = renderWithTheme(
      <ControlledMultiCombobox items={items} searchable allowAdditions />
    );

    const input = getByPlaceholderText('Select manufacturers');

    fireEvent.paste(input, { clipboardData: { getData: () => 'Random Value' } });
    expect(queryAllByRole('tag')).toHaveLength(0);
  });

  it("auto-tokenizes a pasted value that's separated with commas or newlines", async () => {
    const { getByPlaceholderText, queryAllByRole } = renderWithTheme(
      <ControlledMultiCombobox items={items} searchable allowAdditions />
    );

    const input = getByPlaceholderText('Select manufacturers');

    fireEvent.paste(input, {
      clipboardData: { getData: () => 'Random Value 1, Random Value 2 \r\n RandomValue 3' },
    });
    expect(queryAllByRole('tag')).toHaveLength(3);
  });

  it("doesn't allow empty values to be added manually", async () => {
    const { getByPlaceholderText, queryAllByRole } = renderWithTheme(
      <ControlledMultiCombobox items={items} searchable allowAdditions />
    );

    const input = getByPlaceholderText('Select manufacturers');
    fireClickAndMouseEvents(getByPlaceholderText('Select manufacturers'));

    fireEvent.change(input, { target: { value: '  ' } });
    fireEvent.keyDown(input, { key: ',' });

    fireEvent.change(input, { target: { value: '  ' } });
    fireEvent.keyDown(input, { key: 'Enter' });

    fireEvent.blur(input);
    expect(queryAllByRole('tag')).toHaveLength(0);
  });

  it("doesn't allow empty values being pasted", async () => {
    const { getByPlaceholderText, queryAllByRole } = renderWithTheme(
      <ControlledMultiCombobox items={items} searchable allowAdditions />
    );

    const input = getByPlaceholderText('Select manufacturers');

    fireEvent.paste(input, {
      clipboardData: { getData: () => 'Random Value 1\r\n\r\n\r\nRandom Value 2' },
    });
    expect(queryAllByRole('tag')).toHaveLength(2);
  });

  it('works with grouped items', async () => {
    const { getByText, getByPlaceholderText } = renderWithTheme(
      <ControlledMultiCombobox
        itemToGroup={i => i.manufacturer}
        itemToString={i => i.value}
        items={[
          { manufacturer: 'Toyota', value: 'Yaris' },
          { manufacturer: 'Toyota', value: 'Auris' },
          { manufacturer: 'Ford', value: 'Focus' },
        ]}
      />
    );

    fireClickAndMouseEvents(getByPlaceholderText('Select manufacturers'));
    const toyotaGroup = getByText('Toyota');
    const fordGroup = getByText('Ford');
    expect(toyotaGroup).toBeInTheDocument();
    expect(fordGroup).toBeInTheDocument();

    expect(getByText('Yaris')).toBeInTheDocument();
    expect(getByText('Auris')).toBeInTheDocument();
    expect(getByText('Focus')).toBeInTheDocument();

    fireClickAndMouseEvents(await getByText('Focus'));
    expect(fordGroup).not.toBeInTheDocument();
    expect(toyotaGroup).toBeInTheDocument();
  });
});
