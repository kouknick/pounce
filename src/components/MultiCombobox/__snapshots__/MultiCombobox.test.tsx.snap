// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiCombobox renders 1`] = `
.pounce-4 {
  position: relative;
}

.pounce-2 {
  min-height: 47px;
  position: relative;
  border: 1px solid;
  -webkit-transition: border-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: border-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  background-color: transparent;
  border-radius: 4px;
  border-color: #3e516b;
}

.pounce-2:hover {
  border-color: #0081c5;
}

.pounce-2:focus-within {
  border-color: #0081c5;
}

.pounce-2:focus-within label {
  font-weight: 500;
  -webkit-transform: translate(6px, 4px) scale(0.65);
  -moz-transform: translate(6px, 4px) scale(0.65);
  -ms-transform: translate(6px, 4px) scale(0.65);
  transform: translate(6px, 4px) scale(0.65);
}

.pounce-0 {
  vertical-align: top;
  width: 100%;
  height: 100%;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 20px;
  padding-bottom: 8px;
  position: static;
  color: #f6f6f6;
  font-size: 0.875rem;
  font-weight: 500;
  background-color: transparent;
  border: 0;
  cursor: pointer;
  top: 0;
  left: 0;
}

.pounce-0::-webkit-input-placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0::-moz-placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0:-ms-input-placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0::placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0:focus::-webkit-input-placeholder {
  opacity: 0.4;
}

.pounce-0:focus::-moz-placeholder {
  opacity: 0.4;
}

.pounce-0:focus:-ms-input-placeholder {
  opacity: 0.4;
}

.pounce-0:focus::placeholder {
  opacity: 0.4;
}

.pounce-0:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 30px #192231 inset;
  -webkit-text-fill-color: #f6f6f6;
  border-radius: 4px;
}

.pounce-1 {
  pointer-events: none;
  font-size: 0.875rem;
  padding-left: 16px;
  padding-right: 16px;
  color: #bdbdbd;
  top: 0;
  left: 0;
  position: absolute;
  transform-origin: center left;
  -webkit-transform: translate(0, 14px) scale(1);
  -moz-transform: translate(0, 14px) scale(1);
  -ms-transform: translate(0, 14px) scale(1);
  transform: translate(0, 14px) scale(1);
  -webkit-transition: -webkit-transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 400;
}

.pounce-3 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  color: currentColor;
  position: absolute;
  pointer-events: none;
  margin-top: auto;
  margin-bottom: auto;
  top: 0;
  bottom: 0;
  right: 16px;
}

<div>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-0-label"
    class="pounce-4"
    role="combobox"
  >
    <div
      class="pounce-4"
    >
      <div
        aria-disabled="false"
        class="pounce-2"
      >
        <input
          aria-autocomplete="list"
          aria-disabled="false"
          aria-invalid="false"
          aria-labelledby="downshift-0-label"
          aria-readonly="true"
          autocomplete="off"
          class="pounce-0"
          id="downshift-0-input"
          placeholder="Select manufacturers"
          readonly=""
          type="text"
          value=""
        />
        <label
          class="pounce-1"
          for="downshift-0-input"
          id="downshift-0-label"
        >
          Choose a car manufacturer
        </label>
      </div>
      <svg
        class="pounce-3"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10L12 15L17 10H7Z"
        />
      </svg>
    </div>
    <ul
      aria-labelledby="downshift-0-label"
      class="pounce-5"
      id="downshift-0-menu"
      role="listbox"
    />
  </div>
</div>
`;

exports[`MultiCombobox renders with variant="solid" 1`] = `
.pounce-4 {
  position: relative;
}

.pounce-0 {
  vertical-align: top;
  width: 100%;
  height: 100%;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 20px;
  padding-bottom: 8px;
  position: static;
  color: #f6f6f6;
  font-size: 0.875rem;
  font-weight: 500;
  background-color: transparent;
  border: 0;
  cursor: pointer;
  top: 0;
  left: 0;
}

.pounce-0::-webkit-input-placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0::-moz-placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0:-ms-input-placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0::placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0:focus::-webkit-input-placeholder {
  opacity: 0.4;
}

.pounce-0:focus::-moz-placeholder {
  opacity: 0.4;
}

.pounce-0:focus:-ms-input-placeholder {
  opacity: 0.4;
}

.pounce-0:focus::placeholder {
  opacity: 0.4;
}

.pounce-0:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 30px #192231 inset;
  -webkit-text-fill-color: #f6f6f6;
  border-radius: 4px;
}

.pounce-1 {
  pointer-events: none;
  font-size: 0.875rem;
  padding-left: 16px;
  padding-right: 16px;
  color: #bdbdbd;
  top: 0;
  left: 0;
  position: absolute;
  transform-origin: center left;
  -webkit-transform: translate(0, 14px) scale(1);
  -moz-transform: translate(0, 14px) scale(1);
  -ms-transform: translate(0, 14px) scale(1);
  transform: translate(0, 14px) scale(1);
  -webkit-transition: -webkit-transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 400;
}

.pounce-3 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  color: currentColor;
  position: absolute;
  pointer-events: none;
  margin-top: auto;
  margin-bottom: auto;
  top: 0;
  bottom: 0;
  right: 16px;
}

.pounce-2 {
  min-height: 47px;
  position: relative;
  border: 1px solid;
  -webkit-transition: border-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: border-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  background-color: #23344e;
  border-radius: 4px;
  border-color: #23344e;
}

.pounce-2:hover {
  border-color: #0081c5;
}

.pounce-2:focus-within {
  border-color: #0081c5;
}

.pounce-2:focus-within label {
  font-weight: 500;
  -webkit-transform: translate(6px, 4px) scale(0.65);
  -moz-transform: translate(6px, 4px) scale(0.65);
  -ms-transform: translate(6px, 4px) scale(0.65);
  transform: translate(6px, 4px) scale(0.65);
}

<div>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-1-label"
    class="pounce-4"
    role="combobox"
  >
    <div
      class="pounce-4"
    >
      <div
        aria-disabled="false"
        class="pounce-2"
      >
        <input
          aria-autocomplete="list"
          aria-disabled="false"
          aria-invalid="false"
          aria-labelledby="downshift-1-label"
          aria-readonly="true"
          autocomplete="off"
          class="pounce-0"
          id="downshift-1-input"
          placeholder="Select manufacturers"
          readonly=""
          type="text"
          value=""
        />
        <label
          class="pounce-1"
          for="downshift-1-input"
          id="downshift-1-label"
        >
          Choose a car manufacturer
        </label>
      </div>
      <svg
        class="pounce-3"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10L12 15L17 10H7Z"
        />
      </svg>
    </div>
    <ul
      aria-labelledby="downshift-1-label"
      class="pounce-5"
      id="downshift-1-menu"
      role="listbox"
    />
  </div>
</div>
`;
